### PR TITLE
Prevent XML external entity attacks (XXE)

### PR DESCRIPTION
**Pull request by marioja**
See https://rules.sonarsource.com/java/RSPEC-2755

This is not related to a JIRA issue but is just so code vulnerabilities that I found using SonarQube.  See the rules URL for all the details.
- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/edu/hm/hafner/analysis/SecureXmlParserFactory.java [hunk 1]
```diff
@@ -81,6 +74,30 @@ public DocumentBuilder createDocumentBuilder() {
...
+    private void setFeature(final DocumentBuilderFactory factory, final String enabledProperty, final boolean status) {
+        try {
+            factory.setFeature(enabledProperty, status);
+        }
+        catch (ParserConfigurationException ignored) {
+            // ignore and continue
+        }
+    }
+
+    private void secureFactory(final DocumentBuilderFactory factory) {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/SecureXmlParserFactory.java:86]: Can you use the same mechanism here as well (inline with array and for loop)? It looks somehow strange that some values are set using an array and some are set in subsequent calls.

File: src/main/java/edu/hm/hafner/analysis/SecureXmlParserFactory.java [hunk 2]
```diff
@@ -57,21 +59,12 @@ public DocumentBuilder createDocumentBuilder() {
...
             factory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+            secureFactory(factory);
             for (String enabledProperty : ENABLED_PROPERTIES) {
-                try {
-                    factory.setFeature(enabledProperty, true);
-                }
-                catch (ParserConfigurationException ignored) {
-                    // ignore and continue
-                }
+                setFeature(factory, enabledProperty, true);
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/SecureXmlParserFactory.java:64]: Thanks for simplifying this!

File: src/main/java/edu/hm/hafner/analysis/SecureXmlParserFactory.java [hunk 3]
```diff
@@ -81,6 +74,30 @@ public DocumentBuilder createDocumentBuilder() {
...
+        catch (ParserConfigurationException ignored) {
+            // ignore and continue
+        }
+    }
+
+    private void secureFactory(final DocumentBuilderFactory factory) {
+        try {
+            factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
+        }
+        catch (IllegalArgumentException e) {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/SecureXmlParserFactory.java:90]: It seems that this attribute is not supported for the SAX parsers. Do you have an example where it is actually valid? In https://cheatsheetseries.owasp.org/cheatsheets/XML_External_Entity_Prevention_Cheat_Sheet.html it is also not referenced. Do you know why SonarQube does mention it but OWASP not? 

Should we rather use it for `XMLInputFactory`?

File: src/main/java/edu/hm/hafner/analysis/SecureXmlParserFactory.java [hunk 4]
```diff
@@ -91,12 +108,29 @@ public SAXParser createSaxParser() {
...
+            return parser;
         }
         catch (ParserConfigurationException | SAXException exception) {
             throw new IllegalArgumentException("Can't create instance of SAXParser", exception);
         }
     }
+    
+    private void secureParser(final SAXParser parser) {
+        try {
+            parser.setProperty(XMLConstants.ACCESS_EXTERNAL_DTD, "");
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/SecureXmlParserFactory.java:122]: See above: Use a loop.

File: src/main/java/edu/hm/hafner/analysis/parser/checkstyle/TopicRule.java [hunk 5]
```diff
@@ -56,7 +58,9 @@ public void end(final String namespace, final String name) throws Exception {
     @SuppressFBWarnings("SECURITY")
     private String extractNodeContent(final Element subsection) throws TransformerException {
         StringWriter content = new StringWriter();
-        Transformer transformer = TransformerFactory.newInstance().newTransformer();
+        TransformerFactory transformerFactory = TransformerFactory.newInstance();
+        secureFactory(transformerFactory);
+        Transformer transformer = transformerFactory.newTransformer();
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/checkstyle/TopicRule.java:63]: Can you please move that code (and the private methods) to the `SecureXmlParserFactory` class as well? Then all of those security related factory methods are in one place?

**Review by uhafner — CHANGES_REQUESTED** [review]
Thanks for adding the additional checks. I am not sure if this can happen in practice but it should not hurt. Can you have a look at `createXmlStreamReader`? Here it should apply as well. 

There is no need to improve the coverage. Just ignore these checks.

File: src/main/java/edu/hm/hafner/analysis/SecureXmlParserFactory.java (hunk 1 above)
Comment (marioja) [reply src/main/java/edu/hm/hafner/analysis/SecureXmlParserFactory.java:86]: Good idea.  I also created a method to set the features.

File: src/main/java/edu/hm/hafner/analysis/parser/checkstyle/TopicRule.java (hunk 5 above)
Comment (marioja) [reply src/main/java/edu/hm/hafner/analysis/parser/checkstyle/TopicRule.java:63]: I did but I had to create a public static method for that.

Comment (uhafner) [reply src/main/java/edu/hm/hafner/analysis/parser/checkstyle/TopicRule.java:63]: What I mean is that you can use it in the same way:
```
Transformer transformer = new SecureXmlParserFactory().createTransformer();

**Review by uhafner — APPROVED** [review]
I fixed the problem with RevAPI in #702.

Thanks for the PR!

**Pull request merged**